### ISSUE TITLE
support for physical devices connected via USB

### ISSUE THREAD

**Issue by GaryStar** [zebscripts/AFK-Daily#144]
If I connect my phone via USB, I should be able to run the script on the device itself instead of using emulator.

- OS: Windows 10 with ubuntu on WSL
- VERSION="18.04.1 LTS (Bionic Beaver)"

Also, please consider using ```bc``` for arithmetic operations as ```$(())``` does not work with floating point numbers
Ex:
```
ubuntu$ echo $((123.00/2))
bash: 123.00/2: syntax error: invalid arithmetic operator (error token is ".00/2")
ubuntu$ echo $(bc <<< "scale=2; 123.00/2")
61.50
```

**Comment by Zebiano** [comment]
Running the script on a physical device should be possible for you. This script started off as a script meant to run on the phone directly and has since evolved into also supporting Bluestacks (or other emulators). I just tested it on my own device (only part of it), and it worked, so I don't know why it did not for you. Please know that having a rooted phone is necessary, as well as installing BusyBox. And keep in mind all the other requirements (for example a `1080x1920` resolution), you can check them out [here](https://github.com/zebscripts/AFK-Daily/wiki/Supported-Devices#personal-android-device).

Regarding the use of `bc` instead of `$(())`, we cannot use `bc` because Android shells do not have `bc` by default, nor with BusyBox. And I also don't think it's a problem in the first place, as we do not work with floating-point numbers and we also do not plan to. If you have a specific example of floating-point numbers in our script, please let me know. Else we'll tackle this topic once we actually need to.

**Comment by GaryStar** [comment]
first, on my device ```adb shell wm size``` is not returning just a simple resolution but two of them:
```
Physical size: 1440x2960
Override size: 1080x1920
```
i manually used ```wm size``` to change the resolution to ```1080x1920``` before starting the script.

also, ```hexdump``` is failing with ```/system/bin/sh: hexdump: can't execute: Permission denied```
for this i had to change the command to ```su -c hexdump -C```

for the ```$(())``` part, is used in ```support_assist.sh``` for network speed calculations so no really a problem for the main script.

maybe you can address the resolution and hexdump differences in a future script update.
thank you

**Comment by Zebiano** [comment]
> `adb shell vm size`

Oh, interesting. Physical devices are also capable of having an `Override` resolution. Thank you for letting me know. We currently have an [open issue](#143) that is related to the resolution checks. Its goal is to make everything resolution-related a bit easier and automatic. I will address your concerns there!

> `hexdump` is failing with `/system/bin/sh: hexdump: can't execute: Permission denied`

This is also a new one for me, I'm yet to run into permission errors with `hexdump`. Could you tell me some more info regarding your device?
1. Is it rooted? I'm guessing it is since you managed to run `su -c hexdump -C`.
2. Do you have BusyBox installed?
3. What device model is it?
4. What OS is it running?

> `$(())` is used in `support_assist.sh`

Right. In that case then, it's not a problem. I'm currently unsure if the returned values may have floating-point numbers or not, but it's been tested before and it worked. I'll give it another look though, just to be sure! Thank you for the heads-up.

Will close this, as the things to do will be made with #143. Feel free to answer my questions still here though!

**Comment by GaryStar** [comment]
thank you for your quick response.

here is the info:
device is rooted with Magisk and i have BusyBox installed
[ro.build.version.release]: [8.0.0]
[ro.product.manufacturer]: [samsung]
[ro.product.model]: [SM-G965F]

**Comment by Zebiano** [comment]
Interesting then, that you get a `permission denied` error with hexdump... I also have a Samsung with Android `8.0.0`, with Magisk and BusyBox, and I don't require running it with `su`.

I'm thinking of the following reasons:
1. You installed BusyBox in a different way than I did, so that making use of hexdump requires elevated privileges.
2. The way the script got called on your device is maybe different to the way in got called on mine.
3. Maybe something related to Magisk automatically allowing root access to the pc that is connected to the device by default.

I cannot think of other possibilities currently. I will test running the command with `su` also on emulators, and if it works, then I may as well keep it as the new "default". Though I don't particularly like this to be honest, as I feel like it'll create other errors on other devices that were working before. Maybe I can implement some sort of dynamic way of deciding to run it with or without `su`, I don't know. I'll have to find out. If you have any other ideas as to why your particular setup requires `su`, please let me know!

**Comment by GaryStar** [comment]
maybe at the start of the script just run ```adb shell hexdump --help``` and check the result for ```permisison denied``` text. if found, set a variable with ```su -c ```. otherwise set this var to an empty string. then, when calling ```hexdump``` just add the contents of this variable in front of it.

i still don't know why i need elevated privileged for any command added by busybox. i checked ```sbin``` folder and all permissions look ok on both folder and sym links created by busybox.

**Comment by Zebiano** [comment]
I've tested running `su -c hexdump -C` on BlueStacks and it worked as expected.

**Comment by Zebiano** [comment]
I'll actually re-open this issue, as there are some things left to implement that are out of scope for #143. I was thinking of the following things:

- [ ] Implement a new flag `-g` that would change the `afk-daily.sh` script to run `hexdump` with `su -c`. Flag letter is a reference to your username @GaryStar. :wink:
- [ ] Update the way we check for OSs for the `date`s to take WSL into consideration.

I already have ideas on how to do both these tasks, just need to find some time for them.

**Comment by Zebiano** [comment]
EDIT: And I'll re-close it as I've realised both these tasks deserve individual issues. The core of the issue here has been solved already, and should therefore not be re-opened in the first place... Sorry about that.
